File: dynet-cpp/bilstm-tagger.cc
#include <vector>
#include <stdexcept>
#include <fstream>
#include <chrono>
#ifdef BOOST_REGEX
  #include <boost/regex.hpp>
  using namespace boost;
#else
  #include <regex>
#endif

#include <dynet/training.h>
#include <dynet/expr.h>
#include <dynet/dict.h>
#include <dynet/lstm.h>

using namespace std;
using namespace std::chrono;
using namespace dynet;

// Read a file where each line is of the form "word1|tag1 word2|tag2 ..."
// Yields pairs of lists of the form < [word1, word2, ...], [tag1, tag2, ...] >
vector<pair<vector<string>, vector<string> > > read(const string & fname) {
  ifstream fh(fname);
  if(!fh) throw std::runtime_error("Could not open file");
  string str;
  regex re("[ |]");
  vector<pair<vector<string>, vector<string> > > sents;
  while(getline(fh, str)) {
    pair<vector<string>,vector<string> > word_tags;
    sregex_token_iterator first{str.begin(), str.end(), re, -1}, last;
    while(first != last) {
      word_tags.first.push_back(*first++);
      assert(first != last);
      word_tags.second.push_back(*first++);
    }
    sents.push_back(word_tags);
  }
  return sents;
}

class BiLSTMTagger {
public:

  BiLSTMTagger(unsigned layers, unsigned wembed_dim, unsigned hidden_dim, unsigned mlp_dim, ParameterCollection & model, Dict & wv, Dict & tv, unordered_map<string,int> & wc)
                        : wv(wv), tv(tv), wc(wc) {
    unsigned nwords = wv.size();
    unsigned ntags  = tv.size();
    word_lookup = model.add_lookup_parameters(nwords, {wembed_dim});

    // MLP on top of biLSTM outputs 100 -> 32 -> ntags
    pH = model.add_parameters({mlp_dim, hidden_dim*2});
    pO = model.add_parameters({ntags, mlp_dim});

    // word-level LSTMs
    fwdRNN = VanillaLSTMBuilder(layers, wembed_dim, hidden_dim, model); // layers, in-dim, out-dim, model
    bwdRNN = VanillaLSTMBuilder(layers, wembed_dim, hidden_dim, model);
  }

  Dict &wv, &tv;
  unordered_map<string,int> & wc;
  LookupParameter word_lookup;
  Parameter pH, pO;
  VanillaLSTMBuilder fwdRNN, bwdRNN;

  // Do word representation
  Expression word_rep(ComputationGraph & cg, const string & w) {
    return lookup(cg, word_lookup, wv.convert(wc[w] > 5 ? w : "<unk>"));
  }

  vector<Expression> build_tagging_graph(ComputationGraph & cg, const vector<string> & words) {
    // parameters -> expressions
    Expression H = parameter(cg, pH);
    Expression O = parameter(cg, pO);

    // initialize the RNNs
    fwdRNN.new_graph(cg);
    bwdRNN.new_graph(cg);

    // get the word vectors. word_rep(...) returns a 128-dim vector expression for each word.
    vector<Expression> wembs(words.size()), fwds(words.size()), bwds(words.size()), fbwds(words.size());
    for(size_t i = 0; i < words.size(); ++i)
      wembs[i] = word_rep(cg, words[i]);

    // feed word vectors into biLSTM
    fwdRNN.start_new_sequence();
    for(size_t i = 0; i < wembs.size(); ++i)
      fwds[i] = fwdRNN.add_input(wembs[i]);
    bwdRNN.start_new_sequence();
    for(size_t i = wembs.size(); i > 0; --i)
      bwds[i-1] = bwdRNN.add_input(wembs[i-1]);

    // Concatenate and MLP
    for(size_t i = 0; i < wembs.size(); ++i)
      fbwds[i] = O * tanh( H * concatenate({fwds[i], bwds[i]}) );

    return fbwds;
  }

  Expression sent_loss(ComputationGraph & cg, vector<string> & words, vector<string> & tags) {
    vector<Expression> exprs = build_tagging_graph(cg, words), errs(words.size());
    for(size_t i = 0; i < tags.size(); ++i)
      errs[i] = pickneglogsoftmax(exprs[i], tv.convert(tags[i]));
    return sum(errs);
  }

  vector<string> tag_sent(vector<string> & words) {
    ComputationGraph cg;
    vector<Expression> exprs = build_tagging_graph(cg, words), errs(words.size());
    vector<string> tags(words.size());
    for(size_t i = 0; i < words.size(); ++i) {
      vector<float> scores = as_vector(exprs[i].value());
      size_t max_id = distance(scores.begin(), max_element(scores.begin(), scores.end()));
      tags[i] = tv.convert(max_id);
    }
    return tags;
  }

};

int main(int argc, char**argv) {

  time_point<system_clock> start = system_clock::now();

  vector<pair<vector<string>, vector<string> > > train = read("data/tags/train.txt");
  vector<pair<vector<string>, vector<string> > > dev = read("data/tags/dev.txt");
  Dict word_voc, tag_voc;
  unordered_map<string, int> word_cnt;
  for(auto & sent : train) {
    for(auto & w : sent.first) {
      word_voc.convert(w);
      word_cnt[w]++;
    }
    for(auto & t : sent.second)
      tag_voc.convert(t);
  }
  tag_voc.freeze();
  word_voc.convert("<unk>"); word_voc.freeze(); word_voc.set_unk("<unk>");

  // DyNet Starts
  dynet::initialize(argc, argv);
  ParameterCollection model;
  AdamTrainer trainer(model);
  trainer.clipping_enabled = false;

  if(argc != 6) {
    cerr << "Usage: " << argv[0] << " WEMBED_SIZE HIDDEN_SIZE MLP_SIZE SPARSE TIMEOUT" << endl;
    return 1;
  }
  int WEMBED_SIZE = atoi(argv[1]);
  int HIDDEN_SIZE = atoi(argv[2]);
  int MLP_SIZE = atoi(argv[3]);
  trainer.sparse_updates_enabled = atoi(argv[4]);
  int TIMEOUT = atoi(argv[5]);

  // Initilaize the tagger
  BiLSTMTagger tagger(1, WEMBED_SIZE, HIDDEN_SIZE, MLP_SIZE, model, word_voc, tag_voc, word_cnt);

  {
    duration<float> fs = (system_clock::now() - start);
    float startup_time = duration_cast<milliseconds>(fs).count() / float(1000);
    cout << "startup time: " << startup_time << endl;
  }

  // Do training
  start = system_clock::now();
  int i = 0, all_tagged = 0, this_words = 0;
  float this_loss = 0.f, all_time = 0.f;
  for(int iter = 0; iter < 100; iter++) {
    shuffle(train.begin(), train.end(), *dynet::rndeng);
    for(auto & s : train) {
      i++;
      if(i % 500 == 0) {
        trainer.status();
        cout << this_loss/this_words << endl;
        all_tagged += this_words;
        this_loss = 0.f;
        this_words = 0;
      }
      if(i % 10000 == 0) {
        duration<float> fs = (system_clock::now() - start);
        all_time += duration_cast<milliseconds>(fs).count() / float(1000);
        int dev_words = 0, dev_good = 0;
        float dev_loss = 0;
        for(auto & sent : dev) {
          vector<string> tags = tagger.tag_sent(sent.first);
          for(size_t j = 0; j < tags.size(); ++j)
            if(tags[j] == sent.second[j])
              dev_good++;
          dev_words += sent.second.size();
        }
        cout << "acc=" << dev_good/float(dev_words) << ", time=" << all_time << ", word_per_sec=" << all_tagged/all_time << endl;
        if(all_time > TIMEOUT)
          exit(0);
        start = system_clock::now();
      }

      ComputationGraph cg;
      Expression loss_exp = tagger.sent_loss(cg, s.first, s.second);
      float my_loss = as_scalar(cg.forward(loss_exp));
      this_loss += my_loss;
      this_words += s.first.size();
      cg.backward(loss_exp);
      trainer.update();
    }
  }
  return 0;
}

File: dynet-py/bilstm-tagger.py
from __future__ import print_function
import time
start = time.time()

from collections import Counter, defaultdict
import random
import sys
import argparse

import dynet as dy
import numpy as np

parser = argparse.ArgumentParser()
parser.add_argument("--dynet_seed", default=0, type=int)
parser.add_argument("--dynet_mem", default=512, type=int)
parser.add_argument('WEMBED_SIZE', type=int, help='embedding size')
parser.add_argument('HIDDEN_SIZE', type=int, help='hidden size')
parser.add_argument('MLP_SIZE', type=int, help='embedding size')
parser.add_argument('SPARSE', type=int, help='sparse update 0/1')
parser.add_argument('TIMEOUT', type=int, help='timeout in seconds')
args = parser.parse_args()

# format of files: each line is "word1|tag2 word2|tag2 ..."
train_file="data/tags/train.txt"
dev_file="data/tags/dev.txt"

class Vocab:
  def __init__(self, w2i=None):
    if w2i is None: w2i = defaultdict(lambda: len(w2i))
    self.w2i = dict(w2i)
    self.i2w = {i:w for w,i in w2i.items()}
  @classmethod
  def from_corpus(cls, corpus):
    w2i = defaultdict(lambda: len(w2i))
    for sent in corpus:
      [w2i[word] for word in sent]
    return Vocab(w2i)

  def size(self): return len(self.w2i.keys())

def read(fname):
  """
  Read a POS-tagged file where each line is of the form "word1|tag2 word2|tag2 ..."
  Yields lists of the form [(word1,tag1), (word2,tag2), ...]
  """
  with open(fname, "r") as fh:
    for line in fh:
      line = line.strip().split()
      sent = [tuple(x.rsplit("|",1)) for x in line]
      yield sent

train=list(read(train_file))
dev=list(read(dev_file))
words=[]
tags=[]
wc=Counter()
for sent in train:
  for w,p in sent:
    words.append(w)
    tags.append(p)
    wc[w]+=1
words.append("_UNK_")

vw = Vocab.from_corpus([words])
vt = Vocab.from_corpus([tags])
UNK = vw.w2i["_UNK_"]

nwords = vw.size()
ntags  = vt.size()
print ("nwords=%r, ntags=%r" % (nwords, ntags))

# DyNet Starts

model = dy.Model()
trainer = dy.AdamTrainer(model)
trainer.set_clip_threshold(-1.0)
trainer.set_sparse_updates(True if args.SPARSE == 1 else False)

WORDS_LOOKUP = model.add_lookup_parameters((nwords, args.WEMBED_SIZE))

# MLP on top of biLSTM outputs 100 -> 32 -> ntags
pH = model.add_parameters((args.MLP_SIZE, args.HIDDEN_SIZE*2))
pO = model.add_parameters((ntags, args.MLP_SIZE))

# word-level LSTMs
fwdRNN = dy.VanillaLSTMBuilder(1, args.WEMBED_SIZE, args.HIDDEN_SIZE, model) # layers, in-dim, out-dim, model
bwdRNN = dy.VanillaLSTMBuilder(1, args.WEMBED_SIZE, args.HIDDEN_SIZE, model)

def word_rep(w):
  widx = vw.w2i[w] if wc[w] > 5 else UNK
  return WORDS_LOOKUP[widx]

def build_tagging_graph(words):
  dy.renew_cg()
  # parameters -> expressions
  H = dy.parameter(pH)
  O = dy.parameter(pO)

  # initialize the RNNs
  f_init = fwdRNN.initial_state()
  b_init = bwdRNN.initial_state()

  # get the word vectors. word_rep(...) returns a 128-dim vector expression for each word.
  wembs = []
  for i, w in enumerate(words):
    wembs.append(word_rep(w))

  # feed word vectors into biLSTM
  fw_exps = f_init.transduce(wembs)
  bw_exps = b_init.transduce(reversed(wembs))

  # biLSTM states
  bi_exps = [dy.concatenate([f,b]) for f,b in zip(fw_exps, reversed(bw_exps))]

  # feed each biLSTM state to an MLP
  exps = []
  for x in bi_exps:
    r_t = O*(dy.tanh(H * x))
    exps.append(r_t)

  return exps

def sent_loss_precalc(words, tags, vecs):
  errs = []
  for v,t in zip(vecs,tags):
    tid = vt.w2i[t]
    err = dy.pickneglogsoftmax(v, tid)
    errs.append(err)
  return dy.esum(errs)

def sent_loss(words, tags):
  return sent_loss_precalc(words, tags, build_tagging_graph(words))

def tag_sent_precalc(words, vecs):
  log_probs = [v.npvalue() for v in vecs]
  tags = []
  for prb in log_probs:
    tag = np.argmax(prb)
    tags.append(vt.i2w[tag])
  return zip(words, tags)

def tag_sent(words):
  return tag_sent_precalc(words, build_tagging_graph(words))

print ("startup time: %r" % (time.time() - start))
start = time.time()
i = all_time = dev_time = all_tagged = this_tagged = this_loss = 0
for ITER in range(50):
  random.shuffle(train)
  for s in train:
    i += 1
    if i % 500 == 0:   # print status
      trainer.status()
      print(this_loss / this_tagged, file=sys.stderr)
      all_tagged += this_tagged
      this_loss = this_tagged = 0
      all_time = time.time() - start
    if i % 10000 == 0 or all_time > args.TIMEOUT: # eval on dev
      dev_start = time.time()
      good_sent = bad_sent = good = bad = 0.0
      for sent in dev:
        words = [w for w,t in sent]
        golds = [t for w,t in sent]
        tags = [t for w,t in tag_sent(words)]
        if tags == golds: good_sent += 1
        else: bad_sent += 1
        for go,gu in zip(golds,tags):
          if go == gu: good += 1
          else: bad += 1
      dev_time += time.time() - dev_start 
      train_time = time.time() - start - dev_time
      print ("tag_acc=%.4f, sent_acc=%.4f, time=%.4f, word_per_sec=%.4f" % (good/(good+bad), good_sent/(good_sent+bad_sent), train_time, all_tagged/train_time))
      if train_time > args.TIMEOUT:
        sys.exit(0)
    # train on sent
    words = [w for w,t in s]
    golds = [t for w,t in s]

    loss_exp = sent_loss(words, golds)
    my_loss = loss_exp.scalar_value()
    this_loss += my_loss;
    this_tagged += len(golds)
    loss_exp.backward()
    trainer.update()
  print("epoch %r finished" % ITER)
  trainer.update_epoch(1.0)
